middle_end/flambda2/from_lambda/closure_conversion_aux.ml

```diff
@@ -246,28 +243,52 @@ module Acc = struct
   let add_code ~code_id ~code t =
     { t with code = Code_id.Map.add code_id code t.code; }
 
-  let add_symbol_to_free_names ~symbol t =
+  let add_free_names free_names t =
+    { t with free_names = Name_occurrences.union free_names t.free_names; }
+
+  let add_name_to_free_names ~name t =
     { t with
-      free_names_of_current_function =
-        Name_occurrences.add_symbol t.free_names_of_current_function
-          symbol Name_mode.normal;
+      free_names =
+        Name_occurrences.add_name t.free_names
+          name Name_mode.normal;
     }
-  let add_closure_var_to_free_names ~closure_var t =
+
+  let add_simple_to_free_names acc simple =
+    Simple.pattern_match simple
+      ~const:(fun _ -> acc)
+      ~name:(fun name ~coercion:_ ->
+          add_name_to_free_names ~name acc)
+
+  let remove_code_id_or_symbol_from_free_names cis t =
     { t with
-      free_names_of_current_function =
-        Name_occurrences.add_closure_var t.free_names_of_current_function
-          closure_var Name_mode.normal;
+      free_names =
+        Name_occurrences.remove_code_id_or_symbol t.free_names cis;
     }
 
-  let add_continuation_occurrence ~cont ~has_traps t =
+  let remove_symbol_from_free_names symbol t =
+    remove_code_id_or_symbol_from_free_names (Symbol symbol) t
+
+  let remove_var_from_free_names var t =
     { t with
-      free_continuations =
-        Name_occurrences.add_continuation t.free_continuations
-          cont ~has_traps;
+      free_names =
+        Name_occurrences.remove_var t.free_names var;
     }
 
+  let remove_continuation_from_free_names cont t =
+    { t with
+      free_names =
+        Name_occurrences.remove_continuation t.free_names cont;
+    }
+
+  let remove_code_id_from_free_names code_id t =
+    remove_code_id_or_symbol_from_free_names (Code_id code_id) t
+
   let with_free_names free_names t =
-    { t with free_names_of_current_function = free_names; }
+    { t with free_names; }
+
+  let eval_branch_free_names t ~f =
+    let t, res = f { t with free_names = Name_occurrences.empty } in
+    t.free_names, t, res
 
   let measure_cost_metrics acc ~f =
```

Comment:
Would you mind resetting the free_names here too ? This would avoid having to wrap each call to `measure_cost_metrics` with `eval_branch_free_names`. There is only one place where `measure_cost_metrics` is called unwrapped, and it only works because `f` resets the free_names explicitly in that case.

Reply:
(You would likely need to return an additional `free_names_of_body` field, and save/restore the free_names in addition to the cost metrics)